### ISSUE TITLE
coverlet.msbuild not working on .net 7

### ISSUE THREAD

**Issue by maricomic** [coverlet-coverage/coverlet#1391]
Coverlet.msbuild (3.1.2) not working on .net 7

dotnet test /p:CollectCoverage=true /p:CoverletOutputFormat=opencover /p:MergeWith="../CoverageResults/coverage.json"

![image](https://user-images.githubusercontent.com/22876505/195946118-5f7c79a7-7090-4044-8c00-6b4784641fe3.png)

Version .net: 7.0.100-rc.2.22477.23

**Comment by aburgett87** [comment]
Hey @maricomic 
this is an open issue .net 7.0-rc.1 and later. See https://github.com/microsoft/vstest/issues/4014. From the commentary on various issues this will be fixed before the release.

The workaround is to use environment variables.

**Comment by thinkbeforecoding** [comment]
I have this problem with the net7.0 release.
What is the new official way to pass msbuild parameters for tests ?

**Comment by azchohfi** [comment]
Also hitting this.

**Comment by GeorgDangl** [comment]
Hitting the same, just tried to upgrade a library project to .NET 7 and it failed to generate coverage.

We're using https://github.com/nuke-build/nuke, so it was actually quite easy to resolve, but it looks really bad:

```csharp
DotNetTest(c => c
    .EnableCollectCoverage()
    // See https://github.com/coverlet-coverage/coverlet/issues/1391, MSBuild parameters
    // in .NET 7 need to be set via environment variables
    .SetProcessEnvironmentVariable("CollectCoverage", "true")
    .SetCoverletOutputFormat(CoverletOutputFormat.cobertura)
    .SetProcessEnvironmentVariable("CoverletOutputFormat", "cobertura")
```

**Comment by andmos** [comment]
I'm also seeing this - using 
```
dotnet test BikeshareClient/TestBikeshareClient/TestBikeshareClient.csproj /p:CollectCoverage=true /p:Threshold=80 /p:ThresholdType=line /p:CoverletOutputFormat=opencover /p:ExcludeByAttribute=ExcludeFromCodeCoverageAttribute
```
With the msbuild plugin. No coverage shows up, just the standard output:
```
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:    98, Skipped:     0, Total:    98, Duration: 14 s
```

**Comment by jmoralesv** [comment]
Hi all, I'm reporting the same, with the .NET 7 SDK released early this week and using Azure Pipelines and coverlet.msbuild version 3.2.0. 

`dotnet test --configuration $(BuildConfiguration) /p:CollectCoverage=true /p:CoverletOutputFormat=cobertura`

I can confirm that using environment variables in Azure Pipelines works, by using the PowerShell task: https://learn.microsoft.com/en-us/azure/devops/pipelines/process/set-variables-scripts?view=azure-devops&tabs=powershell

Thanks @GeorgDangl for the suggestion!

As I have several projects already configured for coverlet.msbuild, I'll wait until there a solution for this issue with .NET 7.

**Comment by mtelle** [comment]
Hi, for the people could not wait for release 7.0.101. I was able to build the fixed version of dotnet.dll.
You can download from: [Here](https://github.com/mtelle/donetsdk7fixed/raw/main/dotnet.dll)
Just copy it in C:\Program Files\dotnet\sdk\7.0.100 and now it will work.

**Comment by mu88** [comment]
Another option is to add the necessary properties to `csproj`/`Directory.Build.props`, e. g.:
```xml
<PropertyGroup Condition="$(MSBuildProjectName.Contains('.Tests.'))">
  <CollectCoverage>true</CollectCoverage>
  <ExcludeByAttribute>GeneratedCodeAttribute</ExcludeByAttribute>
  <Exclude>[*]*.Migrations.*</Exclude>
  <ExcludeByFile>**/MyFile.cs</ExcludeByFile>
  <CoverletOutputFormat>opencover</CoverletOutputFormat>
</PropertyGroup>
```

**Comment by RayWangQvQ** [comment]
The workaround of environment variables works fine:
```
$env:CollectCoverage='true'
$env:CoverletOutputFormat='cobertura'
$env:CoverletOutput='../'
dotnet test
```

```
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:   464, Skipped:     0, Total:   464, Duration: 5 s - xxx.dll (net6.0)

Calculating coverage result...
  Generating report '..\coverage.cobertura.xml'

+----------------------------+--------+--------+--------+
| Module                     | Line   | Branch | Method |
+----------------------------+--------+--------+--------+
| xxx                    | 98.52% | 94.73% | 98.64% |
+----------------------------+--------+--------+--------+
| xxx                    | 98.97% | 90.16% | 100%   |
+----------------------------+--------+--------+--------+

+---------+--------+--------+--------+
|         | Line   | Branch | Method |
+---------+--------+--------+--------+
| Total   | 98.61% | 94.17% | 99.02% |
+---------+--------+--------+--------+
| Average | 98.74% | 92.44% | 99.32% |
+---------+--------+--------+--------+
```

**Comment by Wolffy101** [comment]
```bash
dotnet test -p:CollectCoverage=true -e:CoverletOutputFormat='lcov' -e:CoverletOutput='./bin/lcov.info'
```

Calculating coverage result...
  Generating report '.\bin\lcov.info'
 
+---------+--------+--------+--------+
|                | Line   | Branch | Method |
+---------+--------+--------+--------+
| Total       | 34.85% | 17.28% | 38.65% |
+---------+--------+--------+--------+
| Average | 34.85% | 17.28% | 38.65% |
+---------+--------+--------+--------+

**Comment by shahidain** [comment]
thanks buddy!

**Comment by jmoralesv** [comment]
Hi all,
I'm writing to report that, after upgrading to version 7.0.101 of the .NET 7 SDK which was released yesterday, I was able to remove the workaround (using environment variables to set the coverlet parameters) and get the coverlet test results with the parameters specified in dotnet test.

`dotnet test --configuration $(BuildConfiguration) /p:CollectCoverage=true /p:CoverletOutputFormat=cobertura`

I tested in my Azure DevOps pipelines and works as expected.

Now I can roll the upgrade to .NET 7 to all my personal projects 🎉

**Comment by samsmithnz** [comment]
Is there any disadvantage to keeping the -e parameters to set the environment variables? I've updated 40+ projects...

**Comment by jmoralesv** [comment]
> Is there any disadvantage to keeping the -e parameters to set the environment variables? I've updated 40+ projects...

I don't think there is any disadvantage, the only question is whether the environment variables approach has precedence over the ones passed as parameters of dotnet test.

**Comment by maryammadzadeh** [comment]
Thanks! Works for me as is now!

**Comment by daveMueller** [comment]
Hi all, the issue was fixed by the microsoft guys with sdk 7.0.101 that was released last week microsoft/vstest#4014. I just checked it with a repro from another related issue and it works fine. Please give it a try again.

**Comment by daveMueller** [comment]
I close this as the external issue is fixed. Feel free to reopen.
